src/gflownet/algo/trajectory_balance.py

```diff
@@ -125,6 +132,47 @@ def create_training_data_from_graphs(self, graphs):
         """
         return [{'traj': generate_forward_trajectory(i)} for i in graphs]
 
+    def get_idempotent_actions(self, g: Graph, gd: gd.Data, gp: Graph, action: GraphAction):
+        """Returns the list of idempotent actions for a given transition.
+
+        Note, this is slow! Correcting for idempotency is needed to estimate p(x) correctly, but
+        isn't generally necessary if we mostly care about sampling approximately from the modes
+        of p(x).
+
+        Parameters
+        ----------
+        g: Graph
+            The state graph
+        gd: gd.Data
+            The Data instance corresponding to g
+        gp: Graph
+            The next state's graph
+        action: GraphAction
+            Action leading from g to gp
+
+        Returns
+        -------
+        actions: List[Tuple[int,int,int]]
+            The list of idempotent actions that all lead from g to gp.
+
+        """
+        iaction = self.ctx.GraphAction_to_aidx(gd, action)
+        if action.action == GraphActionType.Stop:
+            return [iaction]
+        mask_name = self.ctx.action_mask_names[iaction[0]]
+        lmask = getattr(gd, mask_name)
+        nz = lmask.nonzero()
+        actions = [iaction]
+        for i in nz:
+            aidx = (iaction[0], i[0].item(), i[1].item())
```

Comment:
TODO (in the future): create class for this tuple with `__equals__` implementation